azavea/raster-foundry
# Add buckets and scenes API endpoints

The buckets API endpoint supports the following filters:
    - min/max create/modify times
    - create/modified by users
    - organization(s)

The buckets API endpoint supports sorting on the following attributes:
    - create/modify time
    - organization
    - name
    - visibility

The scenes API endpoint supports the following filters:
    - min/max create/modify time
    - acquisition month(s)
    - min/max cloud cover
    - min/max sun azumuth
    - min/max sun elevation
    - organization(s)
    - datasource(s)

The scenes API endpoint supports sorting on the following attributes:
    - create/modify time
    - organization
    - datasource
    - month
    - acquisition time
    - sun azimuth
    - sun elevation
    - cloud cover


### Testing

1. Ensure that all migrations have been run
2. Create scenes with the following attributes
   (must obtain JWT via developer tools after signing in)

   ```json
{
  "thumbnailStatus": "PROCESSING",
  "tags": ["PR Test"],
  "ingestSizeBytes": 0,
  "acquisitionDate": "2016-09-15T12:49:41.324Z",
  "cloudCover": 10.800000190734863,
  "resolutionMeters": 20.0,
  "boundaryStatus": "PROCESSING",
  "sceneMetadata": {
    "instrument_type": "unladen swallow",
    "splines reticulated": 1
  },
  "status": "PROCESSING",
  "organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
  "modifiedBy": "default",
  "visibility": "PUBLIC",
  "datasource": "NASA"
}
   ```

    ```json
{
  "thumbnailStatus": "PROCESSING",
  "tags": ["PR Test Two"],
  "ingestSizeBytes": 0,
  "acquisitionDate": "2016-08-15T12:49:41.324Z",
  "cloudCover": 99,
  "resolutionMeters": 01,
  "boundaryStatus": "PROCESSING",
  "sceneMetadata": {
    "instrument_type": "jason bourne",
    "splines reticulated": 20
  },
  "status": "PROCESSING",
  "organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
  "modifiedBy": "default",
  "visibility": "PRIVATE",
  "datasource": "CIA"
}
    ```
3. List scenes (copy both IDs to a scratch space -- you will need them later)
   : `http://localhost:9000/api/scenes/`

4. Filter scenes and check the following:
   - `http://localhost:9000/api/scenes/?datasource=CIA` should return one
   - `http://localhost:9000/api/scenes/?datasource=CIA&datasource=NASA` should return two
   - any others you want to test

5. Sort scenes and check that they are sorted correctly
   - `http://localhost:9000/api/scenes/?sort=datasource,desc`: NASA should come first
   - any other sorting you want to test

6. Create a bucket with the following json

    ```json
{
  "organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
  "name": "PR Test",
  "description": "This is a test bucket",
  "visibility": "PUBLIC",
  "tags": ["PR Test"]
}
    ```

7. List buckets and ensure it was created

8. Add scenes to bucket via a `POST` request with an empty payload
   using the following URL structure: `http://localhost:9000/api/buckets/<bucketId>/scenes/<sceneId>/`

9. List scenes from bucket: `http://localhost:9000/api/buckets/<bucketId>/scenes/`
    - Should be paginated
    - Should have 2

10. List scenes from bucket with filters using `http://localhost:9000/api/buckets/<bucketId>/scenes/` as the base URL

11. List scenes from bucket with sorting applied to scenes using `http://localhost:9000/api/buckets/<bucketId>/scenes/` as the base URL

12. Delete a scene from the bucket via a `DELETE` request using the
    following URL structure: `http://localhost:9000/api/buckets/<bucketId>/scenes/<sceneId>/`

13. List scenes again for a bucket and ensure the deleted scene has been removed: `http://localhost:9000/api/buckets/<bucketId>/scenes/`
